app/ui.js

```diff
@@ -1168,28 +1168,63 @@ const UI = {
 
     toggleFullscreen() {
-        if (document.fullscreenElement || // alternative standard method
-            document.mozFullScreenElement || // currently working methods
-            document.webkitFullscreenElement ||
-            document.msFullscreenElement) {
-            if (document.exitFullscreen) {
-                document.exitFullscreen();
-            } else if (document.mozCancelFullScreen) {
-                document.mozCancelFullScreen();
-            } else if (document.webkitExitFullscreen) {
-                document.webkitExitFullscreen();
-            } else if (document.msExitFullscreen) {
-                document.msExitFullscreen();
+        // Determine the document using fullscreen. This may either be
+        // just the "document", but if using the viewer in an iframe, you need
+        // to use the parent window's "document" instead.
+        let doc = document;
+        if (window.self !== window.top) {
+            doc = window.parent.document;
+        }
+        // Check the fullscreen status using the correct document as
+        // a reference. The same document is then used to exit fullscreen.
+        if (doc.fullscreenElement || // alternative standard method
+            doc.mozFullScreenElement || // currently working methods
+            doc.webkitFullscreenElement ||
+            doc.msFullscreenElement) {
+            if (doc.exitFullscreen) {
+                doc.exitFullscreen();
+            } else if (doc.mozCancelFullScreen) {
+                doc.mozCancelFullScreen();
+            } else if (doc.webkitExitFullscreen) {
+                doc.webkitExitFullscreen();
+            } else if (doc.msExitFullscreen) {
+                doc.msExitFullscreen();
             }
         } else {
-            if (document.documentElement.requestFullscreen) {
-                document.documentElement.requestFullscreen();
-            } else if (document.documentElement.mozRequestFullScreen) {
-                document.documentElement.mozRequestFullScreen();
-            } else if (document.documentElement.webkitRequestFullscreen) {
-                document.documentElement.webkitRequestFullscreen(Element.ALLOW_KEYBOARD_INPUT);
-            } else if (document.body.msRequestFullscreen) {
-                document.body.msRequestFullscreen();
+            // To activate fullscreen, we need to find the correct document
+            // element, which is usually "document.documentElement". But when
+            // using the viewer in an iframe, this is actually the iframe
+            // element itself in the parent document.
+            let doc_el = document.documentElement;
+            if (window.self !== window.top) {
+                // Seek out the correct iframe from the parent document.
+                // This will produce errors if the iframe does not come from
+                // the same origin, so it is advisable to put the viewer in
+                // iframes only if you are on the same server.
```

Comment:
What kind of errors? Would it be possible to do a try-catch and write a helpful error message in these cases?

Reply:
Thanks for reminding me of that thing, I'm sorry I didn't fix this earlier. The comment was there as a reminder to myself that inspecting iframes just like that might not sit well with some browsers. I just put it there and forgot about it out of laziness.

The test is simple: I added an `<iframe src="https://www.perfact.de">` (shameless plug for my company) before any viewer iframes, so it would come first in the loop, forcing the browser to evaluate "contentDocument" from a foreign source.

Actually, Firefox and Chrome don't care that much (yet), but Edge and IE both give an error.

The fix for that is to wrap only the incriminating "contentDocument" lookup into a "try" block and ignoring any iframes we don't have permission for.

I corrected the comments accordingly and repeated all tests. Now I'm waiting for the travis run, and then I think we're good.

That is, if you don't find other bits where I've been sloppy…